Add exportKubeConfig.additionalSecrets config

nprokopic opened the pull request:
**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind feature

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves ENG-5486


**Please provide a short message that should be published in the vcluster release notes**
Add `exportKubeConfig.additionalSecrets` config and deprecate `exportKubeConfig.secret`.


**What else do we need to know?** 
You can set either old `exportKubeConfig.secret` or the new `exportKubeConfig.additionalSecrets`.

All previous usages of `exportKubeConfig.Secret` have been removed and the only two places where the `exportKubeConfig.Secret` property is now used are:
- `GetAdditionalSecrets()` func in `config/config.go` which returns either old `exportKubeConfig.Secret` or the new `exportKubeConfig.AdditionalSecrets`
- `validateExportKubeConfig(...)` func in `pkg/config/validation.go` where we verify that both old `exportKubeConfig.Secret` and new ``exportKubeConfig.AdditionalSecrets` are not set at the same time

This makes it easy to later entirely remove `exportKubeConfig.Secret`.

**TODOs**

- [x] Add new `exportKubeConfig.additionalSecrets` config
- [x] Deprecate old `exportKubeConfig.secret` config in favor of new `exportKubeConfig.additionalSecrets`
- [x] Update the implementation for exporting kubeconfig.
  - [x] When `exportKubeConfig.secret` is set, `exportKubeConfig.additionalSecrets` is ignored. With this behavior the backward compatibility is preserved.
  - [x] When `exportKubeConfig.additionalSecrets` is set, `exportKubeConfig.secret` is ignored.
  - [x] Block setting both config properties at the same time
- [x] Testing
  - [x] Manually test the changes
  - [x] Ensure that existing tests are passing
  - [x] Add tests for the new config
    - [x] Add unit tests for `exportKubeConfig` config validation
    - [x] Add unit tests for creating exported kubeconfig
- [x] Update code in other places where `exportKubeConfig.secret` is used.
- [x] Other minor improvements to the PR

nprokopic commented on pkg/setup/controllers.go:
```diff
@@ -227,21 +227,21 @@ func SyncKubernetesService(ctx *synccontext.ControllerContext) error {
 
-func WriteKubeConfigToSecret(ctx context.Context, virtualConfig *rest.Config, currentNamespace string, currentNamespaceClient client.Client, options *config.VirtualClusterConfig, syncerConfig *clientcmdapi.Config) error {
-	syncerConfig, err := CreateVClusterKubeConfig(syncerConfig, options)
+func CreateVClusterKubeConfigForExport(ctx context.Context, virtualConfig *rest.Config, syncerConfig *clientcmdapi.Config, options CreateKubeConfigOptions) (*clientcmdapi.Config, error) {
```
New function that is called for both the default secret and the additional secrets (`WriteKubeConfigToSecret` calls it).

Code here is mostly from `WriteKubeConfigToSecret`, and refactored a bit so it can be used for both the default kubeconfig secret and the additional secrets.

To make the implementation even better, this func can be probably moved into / merged with [`CreateVClusterKubeConfig` from `controller_context.go`](https://github.com/loft-sh/vcluster/blob/a4bdf0beed4e5f85f1e07050cb288fb082f614f3/pkg/setup/controller_context.go#L238) (with some `exportKubeConfig` behavior being optional), so that code that creates kubeconfig is in a single place. This would probably make it cleaner and less confusing, as now kubeconfig is created in one func, and then some fields are overridden in another func.

nprokopic commented:
CI errors are most probably not caused by my changes, as I see those same errors in https://github.com/loft-sh/vcluster/pull/2537:
- https://github.com/loft-sh/vcluster/actions/runs/13523283715/job/37787401063?pr=2537
- https://github.com/loft-sh/vcluster/actions/runs/13523283690/job/37787402873?pr=2537

nprokopic commented:
I didn't find any tests related to exporting kubeconfig Secrets, so will see where to add some.

In any case, I've tested it manually, fixed all issues that I found along the way, so this can be reviewed now.

nprokopic commented:
> I didn't find any tests related to exporting kubeconfig Secrets, so will see where to add some.

Added validation tests for `exportKubeConfig`.

johannesfrey commented on .golangci.yml:
```diff
@@ -108,3 +108,6 @@ issues:
     - licenses
+
+  exclude:
+    - "SA1019: (.+)Secret is deprecated: Use AdditionalSecrets instead."
```
Can you add a comment why this is necessary?

nprokopic replied:
Yup, will add a comment line above this.

It's because the linter fails, since we are still using the deprecated field in a couple of places.

nprokopic replied:
Added the comment and updated exclude rule to include only the 2 files where the deprecated field is actually used.

This way if somebody accidentally uses the deprecated Secret field in some other part of code, they will get the linter error.

johannesfrey commented on pkg/config/validation.go:
```diff
@@ -686,6 +692,25 @@ func validateFromHostSyncMappingObjectName(objRef []string, resourceNamePlural s
 	return nil
 }
 
+const (
+	exportKubeConfigBothSecretAndAdditionalSecretsSetError  = "exportKubeConfig.Secret and exportKubeConfig.AdditionalSecrets cannot be set at the same time"
+	exportKubeConfigAdditionalSecretWithoutNameAndNamespace = "additional secret must have name and/or namespace set"
+)
```
nit: should we directly define those as error vars?
```suggestion
var (
	errExportKubeConfigBothSecretAndAdditionalSecretsSet = errors.New("exportKubeConfig.Secret and exportKubeConfig.AdditionalSecrets cannot be set at the same time")
	errExportKubeConfigAdditionalSecretWithoutNameAndNamespace = errors.New("additional secret must have name and/or namespace set")
)
```

nprokopic replied:
Yeah, we can. It will just require also a small test change, as the test is using these consts.

johannesfrey reviewed (commented):
Awesome work! Also how you handle the migration path during the deprecation phase. Only some nits from my side.

nprokopic commented:
> Awesome work! Also how you handle the migration path during the deprecation phase. Only some nits from my side.

Thanks for the feedback! I believe I have addressed the comments you had, PTAL 🙏

johannesfrey reviewed (approved):
lgtm 🎉

lizardruss commented on config/legacyconfig/migrate.go:
```diff
@@ -1048,6 +1059,69 @@ func migrateFlag(key, value string, newConfig *config.Config) error {
+
+	return nil
+}
+
```
Since there is an existing [test](https://github.com/loft-sh/vcluster/blob/a4bdf0beed4e5f85f1e07050cb288fb082f614f3/config/legacyconfig/migrate_test.go) for `MigrateLegacyConfig`, could you add some test cases there to ensure the conversion is as expected?

nprokopic replied:
Good point, will do that.

nprokopic replied:
Added new test cases that include extraArgs being migrated into exportKubeConfig.

lizardruss reviewed (changes requested):
Nice work! I'm requesting some migration test cases, since this is reused by the platform when upgrading a vcluster, and may cause confusion if the somehow the results are not as expected.

nprokopic commented:
Unauthenticated pull request issue again :unamused: Hopefully this gets merged soon https://github.com/loft-sh/vcluster/pull/2556.

Will retry e2e tests here.

nprokopic commented:
@loft-sh/eng-qa got added for the review because of this change in the test https://github.com/loft-sh/vcluster/pull/2542/commits/f8c0808555e1f8c2c094516b10e52a30caf4adc9

Namespace deletion fails (timeout error) because there are still some items in the namespace (I think it's the curl pod). So a made a change to allow the namespace deletion to eventually succeed, in order to make the test a bit more resilient here.

This is the test failure:
```
2025-03-05T23:46:45.5703911Z ------------------------------
2025-03-05T23:46:45.5704869Z CoreDNS resolves host names correctly Test nodes (fake) kubelet is reachable via node hostname
2025-03-05T23:46:45.5707025Z /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:57
2025-03-05T23:47:22.6750158Z   [FAILED] in [AfterEach] - /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:43 @ 03/05/25 23:47:22.674
2025-03-05T23:47:22.6751666Z • [FAILED] [37.105 seconds]
2025-03-05T23:47:22.6756978Z CoreDNS resolves host names correctly [AfterEach] Test nodes (fake) kubelet is reachable via node hostname
2025-03-05T23:47:22.6758277Z   [AfterEach] /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:40
2025-03-05T23:47:22.6759378Z   [It] /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:57
2025-03-05T23:47:22.6760044Z 
2025-03-05T23:47:22.6760270Z   [FAILED] Unexpected error:
2025-03-05T23:47:22.6760694Z       <*url.Error | 0xc000e5dad0>: 
2025-03-05T23:47:22.6761774Z       Delete "https://127.0.0.1:14550/api/v1/namespaces/e2e-coredns-2-tcxlh?timeout=32s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2025-03-05T23:47:22.6762894Z       {
2025-03-05T23:47:22.6763188Z           Op: "Delete",
2025-03-05T23:47:22.6763798Z           URL: "https://127.0.0.1:14550/api/v1/namespaces/e2e-coredns-2-tcxlh?timeout=32s",
2025-03-05T23:47:22.6764524Z           Err: <*http.timeoutError | 0xc000e29530>{
2025-03-05T23:47:22.6765232Z               err: "net/http: request canceled (Client.Timeout exceeded while awaiting headers)",
2025-03-05T23:47:22.6766054Z           },
2025-03-05T23:47:22.6766229Z       }
2025-03-05T23:47:22.6766463Z   occurred
2025-03-05T23:47:22.6767130Z   In [AfterEach] at: /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:43 @ 03/05/25 23:47:22.674
```

And this is one of the errors from the test logs which shows why the namespace deletion failed:
```
INFO	commandwriter/commandwriter.go:128	Unhandled Error	{"component": "vcluster", "component": "k3s", "location": "namespace_controller.go:164", "err": "deletion of namespace e2e-coredns-2-tcxlh failed: unexpected items still remain in namespace: e2e-coredns-2-tcxlh for gvr: /v1, Resource=pods", "logger": "UnhandledError"}
```

lizardruss commented:
> @loft-sh/eng-qa got added for the review because of this change in the test [f8c0808](https://github.com/loft-sh/vcluster/commit/f8c0808555e1f8c2c094516b10e52a30caf4adc9)
> 
> Namespace deletion fails (timeout error) because there are still some items in the namespace (I think it's the curl pod). So a made a change to allow the namespace deletion to eventually succeed, in order to make the test a bit more resilient here.
> 
> This is the test failure:
> 
> ```
> 2025-03-05T23:46:45.5703911Z ------------------------------
> 2025-03-05T23:46:45.5704869Z CoreDNS resolves host names correctly Test nodes (fake) kubelet is reachable via node hostname
> 2025-03-05T23:46:45.5707025Z /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:57
> 2025-03-05T23:47:22.6750158Z   [FAILED] in [AfterEach] - /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:43 @ 03/05/25 23:47:22.674
> 2025-03-05T23:47:22.6751666Z • [FAILED] [37.105 seconds]
> 2025-03-05T23:47:22.6756978Z CoreDNS resolves host names correctly [AfterEach] Test nodes (fake) kubelet is reachable via node hostname
> 2025-03-05T23:47:22.6758277Z   [AfterEach] /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:40
> 2025-03-05T23:47:22.6759378Z   [It] /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:57
> 2025-03-05T23:47:22.6760044Z 
> 2025-03-05T23:47:22.6760270Z   [FAILED] Unexpected error:
> 2025-03-05T23:47:22.6760694Z       <*url.Error | 0xc000e5dad0>: 
> 2025-03-05T23:47:22.6761774Z       Delete "https://127.0.0.1:14550/api/v1/namespaces/e2e-coredns-2-tcxlh?timeout=32s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
> 2025-03-05T23:47:22.6762894Z       {
> 2025-03-05T23:47:22.6763188Z           Op: "Delete",
> 2025-03-05T23:47:22.6763798Z           URL: "https://127.0.0.1:14550/api/v1/namespaces/e2e-coredns-2-tcxlh?timeout=32s",
> 2025-03-05T23:47:22.6764524Z           Err: <*http.timeoutError | 0xc000e29530>{
> 2025-03-05T23:47:22.6765232Z               err: "net/http: request canceled (Client.Timeout exceeded while awaiting headers)",
> 2025-03-05T23:47:22.6766054Z           },
> 2025-03-05T23:47:22.6766229Z       }
> 2025-03-05T23:47:22.6766463Z   occurred
> 2025-03-05T23:47:22.6767130Z   In [AfterEach] at: /home/runner/work/vcluster/vcluster/test/e2e/coredns/coredns.go:43 @ 03/05/25 23:47:22.674
> ```
> 
> And this is one of the errors from the test logs which shows why the namespace deletion failed:
> 
> ```
> INFO	commandwriter/commandwriter.go:128	Unhandled Error	{"component": "vcluster", "component": "k3s", "location": "namespace_controller.go:164", "err": "deletion of namespace e2e-coredns-2-tcxlh failed: unexpected items still remain in namespace: e2e-coredns-2-tcxlh for gvr: /v1, Resource=pods", "logger": "UnhandledError"}
> ```

It looks like the delete call failed on the initial request, and not because it was waiting for items to be removed, which is odd.

nprokopic commented:
> It looks like the delete call failed on the initial request, and not because it was waiting for items to be removed, which is odd.

Yeah, I found that to be a bit weird, the delete request fails after the default 32s timeout.

It took some digging through the logs to find this as well:
```
"err": "deletion of namespace e2e-coredns-2-tcxlh failed: unexpected items still remain in namespace: e2e-coredns-2-tcxlh for gvr: /v1, Resource=pods", "logger": "UnhandledError"}
```

So I figured retrying deletion call few times wouldn't hurt in this case.

nprokopic commented on test/framework/util.go:
```diff
@@ -225,7 +225,15 @@ func (f *Framework) WaitForServiceInSyncerCache(serviceName string, ns string) e
 
 func (f *Framework) DeleteTestNamespace(ns string, waitUntilDeleted bool) error {
-	err := f.VClusterClient.CoreV1().Namespaces().Delete(f.Context, ns, metav1.DeleteOptions{})
+	// Always delete in the background. The vCluster client timeout is set to 32 seconds, so deleting
```
@lizardruss here is the change we talked about.

nprokopic commented:
I was seeing client-side rate limiting in more than few tests, so I increased client's QPS and Burst values in e2e tests (and made those configurable via env vars).

This error was happening frequently:
```
Secrets are synced to host and can be used in Pods synced secret can be used as env source for pod
/home/runner/work/vcluster/vcluster/test/e2e/syncer/fromhost/secrets.go:165
  [FAILED] in [It] - /home/runner/work/vcluster/vcluster/test/e2e/syncer/fromhost/secrets.go:204 @ 03/06/25 21:09:55.606
• [FAILED] [60.049 seconds]
Secrets are synced to host and can be used in Pods [It] synced secret can be used as env source for pod
/home/runner/work/vcluster/vcluster/test/e2e/syncer/fromhost/secrets.go:165

  [FAILED] Unexpected error:
      <*fmt.wrapError | 0xc000725080>: 
      client rate limiter Wait returned an error: context deadline exceeded
      {
          msg: "client rate limiter Wait returned an error: context deadline exceeded",
          err: <context.deadlineExceededError>{},
      }
  occurred
  In [It] at: /home/runner/work/vcluster/vcluster/test/e2e/syncer/fromhost/secrets.go:204 @ 03/06/25 21:09:55.606
```

nprokopic commented:
Also the default service account was missing in e2e tests for PVCs, so pod creation was failing (since pod was using the default service account).

Sometimes, if the control plane is under heavier load, it may take a bit longer (minutes) for the default service account to get created, so I added waiting for the default service account creation.

This was the failure:
```
2025-03-06T20:27:39.4887367Z [0mPersistent volume synced from host cluster [0m[1mTest pvc provisioned successfully and is synced back to vcluster[0m
2025-03-06T20:27:39.4888924Z [38;5;243m/home/runner/work/vcluster/vcluster/test/e2e/syncer/pvc/pvc.go:47[0m
2025-03-06T20:27:41.1684072Z   [38;5;9m[FAILED][0m in [It] - /home/runner/work/vcluster/vcluster/test/e2e/syncer/pvc/pvc.go:98 [38;5;243m@ 03/06/25 20:27:41.168[0m
2025-03-06T20:27:41.1738967Z [38;5;9m• [FAILED] [1.685 seconds][0m
2025-03-06T20:27:41.1739838Z [0mPersistent volume synced from host cluster [38;5;9m[1m[It] Test pvc provisioned successfully and is synced back to vcluster[0m
2025-03-06T20:27:41.1740681Z [38;5;243m/home/runner/work/vcluster/vcluster/test/e2e/syncer/pvc/pvc.go:47[0m
2025-03-06T20:27:41.1741013Z 
2025-03-06T20:27:41.1741435Z   [38;5;9m[FAILED] Unexpected error:
2025-03-06T20:27:41.1741802Z       <*errors.StatusError | 0xc000f24140>: 
2025-03-06T20:27:41.1742533Z       pods "nginx-test" is forbidden: error looking up service account e2e-syncer-pvc-1-ltsea/default: serviceaccount "default" not found
2025-03-06T20:27:41.1743559Z       {
2025-03-06T20:27:41.1746625Z           ErrStatus: {
2025-03-06T20:27:41.1747038Z               TypeMeta: {Kind: "", APIVersion: ""},
2025-03-06T20:27:41.1747485Z               ListMeta: {
2025-03-06T20:27:41.1747829Z                   SelfLink: "",
2025-03-06T20:27:41.1748099Z                   ResourceVersion: "",
2025-03-06T20:27:41.1748356Z                   Continue: "",
2025-03-06T20:27:41.1748624Z                   RemainingItemCount: nil,
2025-03-06T20:27:41.1748867Z               },
2025-03-06T20:27:41.1749060Z               Status: "Failure",
2025-03-06T20:27:41.1749775Z               Message: "pods \"nginx-test\" is forbidden: error looking up service account e2e-syncer-pvc-1-ltsea/default: serviceaccount \"default\" not found",
2025-03-06T20:27:41.1750363Z               Reason: "Forbidden",
2025-03-06T20:27:41.1751029Z               Details: {Name: "nginx-test", Group: "", Kind: "pods", UID: "", Causes: nil, RetryAfterSeconds: 0},
2025-03-06T20:27:41.1751436Z               Code: 403,
2025-03-06T20:27:41.1751635Z           },
2025-03-06T20:27:41.1751803Z       }
2025-03-06T20:27:41.1752035Z   occurred[0m
2025-03-06T20:27:41.1752636Z   [38;5;9mIn [1m[It][0m[38;5;9m at: [1m/home/runner/work/vcluster/vcluster/test/e2e/syncer/pvc/pvc.go:98[0m [38;5;243m@ 03/06/25 20:27:41.168[0m
```

nprokopic commented:
I've got also this test failure (which I think I saw also before):

```
AdmissionWebhook should be able to deny attaching pod
/home/runner/work/vcluster/vcluster/test/e2e/webhook/webhook.go:100
  STEP: Setting up server cert @ 03/06/25 22:28:15.368
  STEP: Create role binding to let webhook read extension-apiserver-authentication @ 03/06/25 22:28:15.656
  STEP: Deploying the webhook pod @ 03/06/25 22:28:15.663
  STEP: Wait for the deployment to be ready @ 03/06/25 22:28:15.675
  STEP: Deploying the webhook service @ 03/06/25 22:29:15.702
  STEP: Wait for the service to be paired with the endpoint @ 03/06/25 22:29:15.727
  STEP: Registering the webhook via the AdmissionRegistration API @ 03/06/25 22:30:15.728
  STEP: create a pod @ 03/06/25 22:30:15.742
  STEP: 'kubectl attach' the pod, should be denied by the webhook @ 03/06/25 22:30:20.762
22:30:20 info Running '/usr/bin/kubectl --kubeconfig=/tmp/vcluster_e2e_kubeconfig_1543812056 --namespace=e2e-webhook-2-ouqed attach --namespace=e2e-webhook-2-ouqed to-be-attached-pod -i -c=container1'
  [FAILED] in [It] - /home/runner/work/vcluster/vcluster/test/framework/helper.go:62 @ 03/06/25 22:30:50.769
• [FAILED] [155.611 seconds]
AdmissionWebhook [It] should be able to deny attaching pod
/home/runner/work/vcluster/vcluster/test/e2e/webhook/webhook.go:100

  [FAILED] unexpected 'kubectl attach' error message. expected to contain "attaching to pod 'to-be-attached-pod' is not allowed", got "timed out waiting for command /usr/bin/kubectl --kubeconfig=/tmp/vcluster_e2e_kubeconfig_1543812056 --namespace=e2e-webhook-2-ouqed attach --namespace=e2e-webhook-2-ouqed to-be-attached-pod -i -c=container1:\nCommand stdout:\n\nstderr:\nIf you don't see a command prompt, try pressing enter.\n"
  In [It] at: /home/runner/work/vcluster/vcluster/test/framework/helper.go:62 @ 03/06/25 22:30:50.769
```

Re-running tests worked. Didn't try to do any fixes/adjustments in this test, as IMO this PR already got a lot bigger than it was supposed to be :see_no_evil:

nprokopic commented:
@johannesfrey @lizardruss PTAL when you have time :pray: (see above comments for the various e2e test improvements I made in this PR)

johannesfrey reviewed (approved):
lgtm

lizardruss reviewed (approved):
lgtm!
